integtest/spec/single_book_spec.rb

```diff
@@ -0,0 +1,209 @@
+# frozen_string_literal: true
+
+RSpec.describe 'building a single book' do
+  ##
+  # Builds a path to afile in the source.
+  def source_file(file)
+    File.expand_path(file, @src)
+  end
+
+  ##
+  # Build a path to a file in the destination.
+  def dest_file(file)
+    File.expand_path(file, @dest)
+  end
+
+  ##
+  # Write a source file.
+  def write_source(source_path, text)
+    path = source_file source_path
+    File.open(path, 'w:UTF-8') do |f|
+      f.write text
+    end
+    path
+  end
+
+  ##
+  # Init a git repo in root and commit any files in it.
+  def init_repo(root)
+    Dir.chdir root do
+      sh 'git init'
+      sh 'git add .'
+      sh "git commit -m 'init'"
+      # Add an Elastic remote so we get a nice edit url
+      sh 'git remote add elastic git@github.com:elastic/docs.git'
+    end
+  end
+
+  def convert_args(from, to)
+    %W[
+      --doc #{from}
+      --out #{to}
+    ]
+  end
+
+  ##
+  # Context for converting asciidoc files into html. To use this include it in
+  # a context and then define a `make_input` method that initializes the source
+  # files and returns the path of the "index" asciidoc file, preferably
+  # using `write_source`.
+  shared_context 'convert' do
+    include_context 'tmp dirs'
+    before(:context) do
+      from = make_input
+      init_repo(File.expand_path('..', from))
+      # Use popen here instead of capture to keep stdin open to appease the
+      # docker-image-always-removed paranoia in build_docs.pl
+      _stdin, out, wait_thr = Open3.popen2e(
+        '/docs_build/build_docs.pl', '--in_standard_docker',
+        *convert_args(from, @dest)
+      )
+      status = wait_thr.value
+      raise_status cmd, out, status unless status.success?
+
+      out
+    end
+    it 'creates the template hash' do
+      expect(dest_file('template.md5')).to file_exist
+    end
+    it 'creates the css' do
+      expect(dest_file('styles.css')).to file_exist
+    end
+    it 'creates the js' do
+      expect(dest_file('docs.js')).to file_exist
+    end
+  end
+
+  ##
+  # Include in a context named after a page to add an assertion that the page
+  # was created and to create some variables extracted from the page's contents
+  #
+  # body - the non-template body of the page
+  # title - the title of the page
+  shared_context 'page body' do
+    let(:file) do |example|
+      dest_file(example.example_group.description)
```

Comment:
hmm this is a tight and implicit requirement of the `page body` context blocks. Is there a better and more explicit way to capture the argument to `dest_file`?

Reply:
I expect I'll include this many, many times and I liked the convenience of this. I can certainly have something like

```
context 'for index.html' do
  let(:file_name) { 'index.html' }`
end
```

Instead of making a shared context I could do something like this:
```
def page_context(file_name) do
  context "for #{file_name}" do
    let(:file) do |example|
    ....
    yield
  end
end
```

That way it is a bit more explicit that you are making a context for checking on the page.